### DISCUSSION TITLE
Cobbler 3.3.1 cobblerd startup time is immense

### DISCUSSION THREAD

**Discussion by AndyGrill** [cobbler/cobbler#3578] [Q&A]
Hi, is there any way in Cobbler 3.3.1 to skip the cobblerd initialization that performs these steps on every system present in the service?  I obscured the system and profile for privacy reasons, and it shouldn't matter for this discussion.

> [Daemon] 2024-02-02T16:58:18 - DEBUG | Parent of System "<system>" not found. Thus skipping removal from children list.
> [Daemon] 2024-02-02T16:58:18 - DEBUG | Name of System "<system>" was not found in the children of Item "<profile>"

In one of our sites during this conversion with just about 5300 systems (full amount will be about 52K), the startup time for cobbler iterating through each of these systems is astronomical (about 16 MINUTES before cobbler responds to edit/addition/remove requests):

> [Daemon] 2024-02-02T16:33:08 - INFO | Cobbler startup completed  in **967.7444179058075** seconds

Is there any help someone could provide that could slim down the cobblerd startup initialization or is this so deep in the service that there's no way out of this?

**Some background:**

I had previously contacted in this thread: https://github.com/cobbler/cobbler/discussions/2980 regarding optimizations and tweaks for cobbler, as my environment has tens of thousands or even hundreds of thousands of system entries in one or two cases for the cobbler service.  At the time we were still on 2.6.6 with no real upgrades in sight, so the conversation didn't get too far.  Recently, we've been working on upgrading from Cobbler 2.6.6 to 3.3.1 and we've put a tremendous amount of effort into this process.  It wasn't until beginning the pilot for one of our medium-sized sites that I discovered this long initialization time.

I tried picking at the python libraries under /usr/lib/python3.6/site-packages/cobbler/ as we've already had to make some tweaks for bug fixes such as the default interface being created with a blank mac even if an interface name is given, but I've been unable to find the source of this to remove it and allow cobbler to start up faster.  Cobblerd NEEDS to start much faster than it does currently.  Using a stop watch on our biggest site (~500k system entries) in Cobbler 2.6.6, the cobbler startup time is roughly 165 seconds (2 minutes, 45 seconds) which isn't amazing, but is acceptable.  In this new version, I can't imagine how long this would take, probably a day or more to initialize.

**Comment by SchoolGuy** [comment]
My resources to test something like this are simply not present in the community. Cobbler 3.x.x has shifted towards validating the data it loads. The major problem of all installation instances I encountered was large problems with invalid data. This is a problem that Cobbler 3 has tried to solve so far. I really want to speed up the load times but Cobbler 3.3.1 is the wrong version for.

I have created https://github.com/cobbler/cobbler/issues/3341 to gather input for the performance effort. Nobody so far has given any real-world expectations and without a clear goal in mind, it is hard to optimize because in my environment it is acceptable to have longer startup times.

I saw that you want to move to SLES 15 SP4 in the other discussion. The only thing - if you need guarantees - I can offer is that we connect officially via work. I work at SUSE and as such have options here. Company internal the cloud-init support for Cobbler is more important for us as such to override this priority things would need to go through the proper channels.

A new development since our last discussion is that cobbler-tftp should now be operational. I had not a chance to roll it out internally as this was finished literally a week ago but it will remove the need for a `cobbler sync` in case you boot via TFTP and PXE.

I thank you for coming forward publically and stating your problem and I would love to continue this discussion. The way you communicate your expectations is helping a lot.

Edit: If I understand you right then with 500k systems and an acceptable startup time of 2 minutes that means we have 0.2 milliseconds roughly to load the data from disk and validate it. This doesn't include distros and profiles that are the parent items. I will take some time on the way home from FOSDEM to think what can be done. I think using spinning disks makes this kind of IO next to impossible. I have an NVMe-based server at work I will spin up for testing...

Edit 2: It seems this will be quite the challenge because Cobbler is Python. Just loading a single JSON (no validation whatsoever) seems to be too slow... But let's see if I can figure out something....

```
enno@dell-laptop /tmp [1]> time python3 -c "import json; import pathlib; json.loads(pathlib.Path('/tmp/test.json').read_text())"

________________________________________________________
Executed in  100.60 millis    fish           external
   usr time   62.60 millis    1.03 millis   61.57 millis
   sys time   32.65 millis    0.24 millis   32.41 millis
```

**Comment by SchoolGuy** [comment]
Okay since my interest was caught I did some experiments on my local laptop. These are looking like the following:

- Generator Script
- Read Script
- Some shell commands

```python
#!/usr/bin/python3

import json
import pathlib

BASE_FOLDER = pathlib.Path("cobbler-test")
ITERATIONS = 500000

def generate_file(index: int):
    (pathlib.Path(BASE_FOLDER) / f"test{index:06d}.json").write_text(json.dumps({"name": f"testname{index}"}))

BASE_FOLDER.mkdir(exist_ok=True)
for iteration in range(0, ITERATIONS):
    generate_file(iteration)
```

```python
#!/usr/bin/python3

import json
import pathlib
from typing import Dict

BASE_FOLDER = pathlib.Path("cobbler-test")
ITERATIONS = 500000

collection: Dict[str, Dict[str, str]] = {}

def read_file(index: int):
    my_dict = json.loads((pathlib.Path(BASE_FOLDER) / f"test{index:06d}.json").read_text())
    collection[my_dict["name"]] = my_dict

for iteration in range(0, ITERATIONS):
    read_file(iteration)

print(len(collection))
```

```
enno@dell-laptop ~/tmp> time ./../read.py

________________________________________________________
Executed in   23.19 secs    fish           external
   usr time   16.94 secs  981.00 micros   16.94 secs
   sys time    6.19 secs  462.00 micros    6.19 secs

enno@dell-laptop ~/tmp> time ./../read.py
500000

________________________________________________________
Executed in   37.30 secs    fish           external
   usr time   29.41 secs    4.45 millis   29.41 secs
   sys time    6.43 secs    0.52 millis    6.43 secs
enno@dell-laptop ~/tmp [1]> du -hs
2,0G 
```

This means that if Cobbler would run on my Dell Latitude 7480 with my Intel i7-6600U and a Crucial CT1000P2SSD8, I have 90 seconds left for the Cobbler-specific parts and loading a few more bytes. I think this is potentially possible but it will be very hard to achieve.

If I would get 3.4.0/4.0.0 (not final yet what it will be) you could utilize the lazy loading feature that doesn't load everything on startup of the daemon.

**Reply by SchoolGuy** [reply]
Now if we want to validate that the JSON have the correct schema (to save the countless amounts of `isinstance` calls) we would be using the `schema` Python library. If the `read` script is extended we get the following results:

```python
#!/usr/bin/python3

import json
import pathlib
from typing import Dict

import schema

BASE_FOLDER = pathlib.Path("cobbler-test")
ITERATIONS = 500000

my_schema = schema.Schema({"name": str})
collection: Dict[str, Dict[str, str]] = {}

def read_file(index: int):
    my_dict = json.loads((pathlib.Path(BASE_FOLDER) / f"test{index:06d}.json").read_text())
    if not my_schema.is_valid(my_dict):
        print(f"error with index {index}")
    collection[my_dict["name"]] = my_dict

for iteration in range(0, ITERATIONS):
    read_file(iteration)

print(len(collection))

```

Results (same hardware as above):

```
enno@dell-laptop ~/tmp [1]> time ./../read.py
500000

________________________________________________________
Executed in   32.00 secs    fish           external
   usr time   26.20 secs    2.65 millis   26.19 secs
   sys time    5.80 secs    0.00 millis    5.80 secs
enno@dell-laptop ~/tmp> time ./../read.py
500000

________________________________________________________
Executed in   34.13 secs    fish           external
   usr time   27.80 secs    1.77 millis   27.80 secs
   sys time    6.25 secs    0.00 millis    6.25 secs

enno@dell-laptop ~/tmp> time ./../read.py
500000

________________________________________________________
Executed in   33.63 secs    fish           external
   usr time   27.06 secs    1.10 millis   27.06 secs
   sys time    6.56 secs    0.41 millis    6.56 secs

enno@dell-laptop ~/tmp> time ./../read.py
500000

________________________________________________________
Executed in   32.49 secs    fish           external
   usr time   26.80 secs    1.09 millis   26.80 secs
   sys time    5.69 secs    0.40 millis    5.69 secs
```

What is evident is that the performance varies greatly depending on the execution. As such multiple rounds need to be done once I am at work and have a proper server at hand that doesn't run a graphical stack and random userspace background tools.

**Reply by AndyGrill** [reply]
This is great info, thank you very much for taking a look at this already :)  My intel email is andrew.grill@intel.com, I'm unsure how to contact you myself.  If we could please get together sometime, I'd be happy to show you the sorts of stuff we have and demonstrate the slowness I found in our pilot site or use the test to try various options.

**Reply by SchoolGuy** [reply]
Just for documentation: Andy and I had a meeting today and agreed on a few things. I can't promise anything but I have a vision to go forward so that the performance needs of Andy are met.

**Reply by SchoolGuy** [reply]
Just for transparency with the upstream community. This is now handled as a SUSE Package Hub Bug that can be followed at: https://bugzilla.suse.com/show_bug.cgi?id=1219933

**Reply by SchoolGuy** [reply]
3.3.5 is out: https://github.com/cobbler/cobbler/releases/tag/v3.3.5

The release should adequately address the startup time of Cobbler.

**Reply by SchoolGuy** [reply]
3.3.6 is out: https://github.com/cobbler/cobbler/releases/tag/v3.3.6

This addresses the missing backports from openSUSE.
